[faudio] add native feature for Windows

rkitover opened the pull request:
Add `sdl3` feature for Windows to use the SDL3 backend instead of the Windows platform backend. Useful for Windows XP which does not support the Windows platform backend.

<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

<!-- If this PR updates an existing port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

END OF PORT UPDATE CHECKLIST (delete this line) -->

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

rkitover commented:
@dg0yt What do you think, does this make sense?

rkitover commented:
Done, thank you for taking a look!

BillyONeal reviewed (commented):
Reviewing for:

* https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#do-not-use-features-to-implement-alternatives , and
* https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#default-features-should-enable-behaviors-not-apis

The change here controls PLATFORM_WIN32, so that must not trigger visible 'alternative'ness or control presence of APIs.

That changes FAUDIO_WIN32_PLATFORM, HAVE_WMADEC=1, XNASONG=OFF https://github.com/FNA-XNA/FAudio/blob/8da412ed6c78c6e225b873bb12cbc903d5389192/CMakeLists.txt#L122

and inverted controls `find_dependency(SDL2` : https://github.com/FNA-XNA/FAudio/blob/8da412ed6c78c6e225b873bb12cbc903d5389192/cmake/config.cmake.in#L5 which is bogus. Submitted fix https://github.com/FNA-XNA/FAudio/pull/384 but that fix doesn't seem required for the effect this PR is trying to have.

FAUDIO_WIN32_PLATFORM seems to control swapping in and out an entire implementation of the library intended to provide the same interface, so it at least *seems* OK.

HAVE_WMADEC=1 seems problematic, it means turning on the SDL3 feature would remove these APIs: https://github.com/FNA-XNA/FAudio/blob/8da412ed6c78c6e225b873bb12cbc903d5389192/src/FAudio_internal.h#L782-L786

XNASONG seems very featureish / OK.

Tagging vcpkg-team-review. My understanding of the situation is that technically the proposed feature removes APIs (those controlled by `HAVE_WMADEC`). It seems like upstream's intent is for the different backends to be indistinguishable for most customers in most cases, but that different backends can sometimes provide 'fun extras' like this.

On the one hand I can see an argument that those additional bits aren't "meaningful" for purposes of our alternatives analysis. On the other hand the slated use case for ever turning this feature on, Windows XP support, is an explicit non-goal for vcpkg. So I'm torn on whether we can merge this as proposed.

Just always having an SDL3 dependency (that is, not having the feature and making the dependency unconditional) seems like it would be unambiguously OK. Given that we haven't seen anyone other than @rkitover contribute here ( https://github.com/microsoft/vcpkg/commits/master/ports/faudio / https://github.com/microsoft/vcpkg/issues?q=is%3Aissue%20state%3Aopen%20faudio ) if this is the setting they want that's a good argument that we should just make it unconditional.

BillyONeal commented:
> but that fix doesn't seem required for the effect this PR is trying to have

I stand corrected, as written this PR makes `faudio[sdl3]` on Windows customers broken because there is no `find_dependency(SDL3`. I don't think my 10 minute submitted fix needs to be the actual fix but we probably at least need to replace 'SDL2' with 'SDL3' in a patch if we keep this as a separate feature.

BillyONeal committed:
[faudio] add find_dependency fix for SDL

rkitover commented:
Thanks Kai and Billy, I'll get back to this in a day or two and we'll sort this out.

BillyONeal commented:
@rkitover Just to clarify, would you be strongly opposed to just removing the feature and making the dependency unconditional? That way we don't have to engage with the 'turning on this feature removes APIs' problem/question and it sounds like that's how you want to use this library anyway.

rkitover commented:
@BillyONeal Well, the issue is that the `PLATFORM_WIN32` code does not need SDL3 and is a lower latency/better implementation on Windows than using the SDL3 code.

BillyONeal commented:
But that even more seems alternative-ish, right? It suggests someone would need to be able to depend on a feature being *off* which isn't a feature of features :/

rkitover commented:
Hi Billy, it may help if I explain what this is for.

FAudio is an implementation of the Windows XAudio2 APIs, which are usually used in conjunction with DirectX and related APIs.

As I understand it, the API is similar if not exactly the same.

It uses SDL3 or SDL2 as the operating system audio interface on all platforms (macOS and Linux) except Windows, where it also has the better `PLATFORM_WIN32` implementation.

The reasons that someone may want to use it are if they need a game programming type of audio library that is cross-platform. In addition, using the actual XAudio2 APIs can be problematic due to dlls being unavailable for a specific version etc., while FAudio would give a similar quality implementation using the `PLATFORM_WIN32` code.

My overlay uses `mingw && x86` as a target for Windows XP, which is not really correct because this toolchain can also be used to target Vista.

I don't understand this `WMADEC` functionality and APIs. I would guess that they are very unlikely to be needed by the vast majority of users, who are mainly interested in the APIs analogous to XAudio2.

If this `WMADEC` functionality depends on SDL3, I can research implementing it Win32 to use in-conjunction with `PLATFORM_WIN32` upstream.

If "features must be treated as additive functionality", then this `sdl3` feature is not correct, because it's more like a reduction in functionality. 

So, what if we make a `native` feature that would be a default feature and enable `PLATFORM_WIN32` and possibly `PLATFORM_MACOS` in the future or whatever, while `faudio[core]` would select the SDL3 backend? And I will research implementing `WMADEC` natively.

rkitover commented:
I misread your previous post, so `WMADEC` is a part of the `PLATFORM_WIN32` backend and not SDL3.

I don't know if my post was edited to remove my proposal, so I will describe it again.

I suggested making a `native` feature that will be part of `default-features`. It would do nothing on SDL3 platforms until they have a native backend.

In the case of installing `faudio[core]` on a Windows platform, it would depend on and build with the SDL3 backend.

This would make it both a "feature" and "additive."

Users would not have to make any changes to their use of this port.

I can also look into separating out the `WMADEC` functionality for use with the SDL3 backend, either on Windows or on all platforms.

BillyONeal commented:
@vicroms @ras0219-msft and I discussed one more time today.

> I misread your previous post, so `WMADEC` is a part of the `PLATFORM_WIN32` backend and not SDL3.
> 
> I don't know if my post was edited to remove my proposal, so I will describe it again.
> 
> I suggested making a `native` feature that will be part of `default-features`. It would do nothing on SDL3 platforms until they have a native backend.
> 
> In the case of installing `faudio[core]` on a Windows platform, it would depend on and build with the SDL3 backend.
> 
> This would make it both a "feature" and "additive."
> 
> Users would not have to make any changes to their use of this port.
> 
> I can also look into separating out the `WMADEC` functionality for use with the SDL3 backend, either on Windows or on all platforms.

We would be able to merge this proposal, restating:

* `"sdl3"` becomes an unconditional dependency
* `"native"` becomes a feature that replaces the SDL3 backend with the `PLATFORM_WIN32` backend on Windows
* `"native"` is in `"default-features"`

However, we observe that it leaves a dependency on `"sdl3"` even when the `"native"` feature is turned on, which may be undesirable.

We still prefer just making the decision of backend such that it works for your use cases rather than adding a feature, but if you prefer that tradeoff we could merge that.

rkitover commented:
Looks like we also need to port `XNASONG` to `PLATFORM_WIN32`, whatever that is.

rkitover commented:
@dg0yt Is there any way I could express this such that SDL3 is a dependency of `faudio[core]` but not `faudio[native]` on Windows?

rkitover committed:
[faudio] add native default feature for Windows

Add the `native` feature in `default-features` to use the
`PLATFORM_WIN32` backend for Windows targets, which was the default
behavior previously and continues to be.

This allows installing `faudio[core]` to build with the SDL3 backend,
which is useful for Windows XP which does not support the Windows
platform backend.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover commented:
Despite having implemented this this way currently in this PR, I don't really want a useless dependency.

If this can be expressed this way without it, I'd prefer that, otherwise some kind of OS version variable as I mentioned, and if not than your original suggestion.

BillyONeal commented:
/azp run

dg0yt commented:
For the Android "deployment target", there is already project-mode variable `CMAKE_SYSTEM_VERSION`, initialized by the optional triplet variable `VCPKG_CMAKE_SYSTEM_VERSION`.

I don't think there is a need for a similar host variable. (In the native triplet, it is just the same as the target variable.)

I don't see a general place or need for targeting specific linux distros. Is there a convincing use-case in the context of port faudio?

BillyONeal commented:
If we're just talking about this faudio setting, I don't think it's really justified to try to make up a setting to track target operating system version; in particular there is no generally agreed upon way to communicate this between build systems.

Additionally, Windows XP support remains an extremely explicit non-goal for vcpkg. Even if we added this, we would reject attempts to merge ports querying for XP anyway. This was litigated back in 2020 when XP had been out of support for 6 years, and now that it's out of support for 12 years things have not gotten better. https://github.com/microsoft/vcpkg/issues/1015

Our position here remains:

> @vicroms @ras0219-msft and I discussed one more time today.
> 
> > I misread your previous post, so `WMADEC` is a part of the `PLATFORM_WIN32` backend and not SDL3.
> > I don't know if my post was edited to remove my proposal, so I will describe it again.
> > I suggested making a `native` feature that will be part of `default-features`. It would do nothing on SDL3 platforms until they have a native backend.
> > In the case of installing `faudio[core]` on a Windows platform, it would depend on and build with the SDL3 backend.
> > This would make it both a "feature" and "additive."
> > Users would not have to make any changes to their use of this port.
> > I can also look into separating out the `WMADEC` functionality for use with the SDL3 backend, either on Windows or on all platforms.
> 
> We would be able to merge this proposal, restating:
> 
> * `"sdl3"` becomes an unconditional dependency
> * `"native"` becomes a feature that replaces the SDL3 backend with the `PLATFORM_WIN32` backend on Windows
> * `"native"` is in `"default-features"`
> 
> However, we observe that it leaves a dependency on `"sdl3"` even when the `"native"` feature is turned on, which may be undesirable.
> 
> We still prefer just making the decision of backend such that it works for your use cases rather than adding a feature, but if you prefer that tradeoff we could merge that.

Feel free to use an overlay-port if you need this change and are unwilling to consider one of the above alternatives.
